File: server/src/app/start.ts
// Starting a pick: validate the request and geocode it (SPEC.md F1).

import { randomBytes } from "node:crypto";
import { z } from "zod";
import { distinctLocations, normaliseAddress, type Geocoder, type Location } from "../domain/places";
import { DEFAULT_MIN_POPULATION } from "../domain/pool";
import { DEFAULT_MAX_WAIT_MIN, MAX_WAIT_MIN_RANGE } from "../domain/race";
import { newSession, type PickSession } from "../domain/session";
import { GEOCODE_TTL_MS, isFresh, type Store } from "../domain/store";
import type { Iso } from "../domain/time";
import { log } from "../log";

export const DEFAULT_RADIUS_M = 200;
export const RADIUS_MIN = 50;
export const RADIUS_MAX = 2000;

/** The body of `POST /picks` (§5). */
export const startInput = z.object({
  address: z.string().optional(),
  lat: z.number().optional(),
  lon: z.number().optional(),
  radius_m: z.number().int().optional(),
  min_population: z.number().int().nonnegative().optional(),
  include_visited: z.boolean().optional(),
  max_wait_min: z.number().int().optional(),
  /** With lat+lon: the address to show, e.g. a match chosen from `GET /geocode` (F1.2). */
  label: z.string().max(300).optional(),
});
export type StartInput = z.infer<typeof startInput>;

export class InvalidRequest extends Error {}
export class AddressNotFound extends Error {
  constructor() {
    super("address not found");
  }
}

/** More than one place matches; the caller must choose (F1.2). */
export class AmbiguousAddress extends Error {
  constructor(readonly matches: Location[]) {
    super(`${matches.length} places match that address`);
  }
}

/** Validate F1.1 and resolve the location (F1.2–F1.3); returns a new session. */
export async function startPick(
  geocoder: Geocoder,
  cache: Pick<Store, "getGeocode" | "putGeocode">,
  input: StartInput,
  pickId: string,
  now: Iso,
): Promise<PickSession> {
  const radius = input.radius_m ?? DEFAULT_RADIUS_M;
  if (radius < RADIUS_MIN || radius > RADIUS_MAX) {
    throw new InvalidRequest(`radius_m must be ${RADIUS_MIN}–${RADIUS_MAX}`);
  }
  const maxWait = input.max_wait_min ?? DEFAULT_MAX_WAIT_MIN;
  const [minWait, maxWaitLimit] = MAX_WAIT_MIN_RANGE;
  if (maxWait < minWait || maxWait > maxWaitLimit) {
    throw new InvalidRequest(`max_wait_min must be ${minWait}–${maxWaitLimit}`);
  }
  const address = input.address?.trim() || undefined;
  let location: Location;
  if (address !== undefined && input.lat === undefined && input.lon === undefined) {
    const matches = await lookupAddress(geocoder, cache, address, now);
    const [only, ...others] = matches;
    if (!only) throw new AddressNotFound();
    if (others.length > 0) throw new AmbiguousAddress(matches);
    location = only;
  } else if (address === undefined && input.lat !== undefined && input.lon !== undefined) {
    if (Math.abs(input.lat) > 90 || Math.abs(input.lon) > 180)
      throw new InvalidRequest("lat/lon out of range");
    location = {
      lat: input.lat,
      lon: input.lon,
      display_name: input.label?.trim() || `${input.lat.toFixed(5)}, ${input.lon.toFixed(5)}`,
    };
  } else {
    throw new InvalidRequest("give exactly one of address or lat+lon");
  }
  return newSession(
    pickId,
    now,
    {
      radius_m: radius,
      min_population: input.min_population ?? DEFAULT_MIN_POPULATION,
      include_visited: input.include_visited ?? false,
      max_wait_min: maxWait,
    },
    location,
  );
}

/**
 * Distinct places matching `address` (F1.2), best first, cached per address and
 * geocoder scope (F1.3). Matches within SAME_PLACE_M of a better one are merged.
 */
export async function lookupAddress(
  geocoder: Geocoder,
  cache: Pick<Store, "getGeocode" | "putGeocode">,
  address: string,
  now: Iso,
): Promise<Location[]> {
  const key = `${geocoder.scope}|${normaliseAddress(address)}`;
  const cached = await cache.getGeocode(key).catch(() => null);
  if (cached && Array.isArray(cached.results) && isFresh(cached.created_at, GEOCODE_TTL_MS, now)) {
    return cached.results;
  }
  const results = distinctLocations(await geocoder.search(address));
  await cache
    .putGeocode(key, { results, created_at: now })
    .catch((e) => log.warn("geocode cache write failed", { error: String(e) }));
  return results;
}

const CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

/** A new, time-ordered pick id (ULID). */
export function newPickId(nowMs = Date.now()): string {
  let time = "";
  let t = nowMs;
  for (let i = 0; i < 10; i++) {
    time = CROCKFORD[t % 32] + time;
    t = Math.floor(t / 32);
  }
  const rand = [...randomBytes(16)].map((b) => CROCKFORD[b % 32]).join("");
  return time + rand;
}

File: crates/app/src/start.rs
//! Starting a pick: validate the request and geocode it (SPEC.md F1).

use chrono::{DateTime, Utc};
use domain::geo::{Geocoder, Location, normalise_address};
use domain::pool::DEFAULT_MIN_POPULATION;
use domain::session::{PickRequest, PickSession};
use domain::store::{CachedLocation, GEOCODE_TTL, GeocodeCache, is_fresh};
use serde::Deserialize;

pub const DEFAULT_RADIUS_M: u32 = 200;
pub const RADIUS_RANGE: std::ops::RangeInclusive<u32> = 50..=2000;

/// The body of `POST /picks` (§5).
#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
pub struct StartInput {
    pub address: Option<String>,
    pub lat: Option<f64>,
    pub lon: Option<f64>,
    pub radius_m: Option<u32>,
    pub min_population: Option<u64>,
    pub include_visited: Option<bool>,
}

#[derive(Debug, thiserror::Error, PartialEq)]
pub enum StartError {
    #[error("invalid request: {0}")]
    InvalidRequest(String),
    #[error("address not found")]
    AddressNotFound,
    #[error("geocoder unavailable: {0}")]
    GeocoderUnavailable(String),
}

/// Validate F1.1 and resolve the location (F1.2–F1.3). Returns a new session in
/// `finding_race`; the caller stores it and starts the workflow (F1.4).
pub async fn start_pick<G, C>(
    geocoder: &G,
    cache: &C,
    input: StartInput,
    pick_id: String,
    now: DateTime<Utc>,
) -> Result<PickSession, StartError>
where
    G: Geocoder + Sync,
    C: GeocodeCache + Sync,
{
    let radius_m = input.radius_m.unwrap_or(DEFAULT_RADIUS_M);
    if !RADIUS_RANGE.contains(&radius_m) {
        return Err(StartError::InvalidRequest(format!(
            "radius_m must be {}–{}",
            RADIUS_RANGE.start(),
            RADIUS_RANGE.end()
        )));
    }
    let request = PickRequest {
        radius_m,
        min_population: input.min_population.unwrap_or(DEFAULT_MIN_POPULATION),
        include_visited: input.include_visited.unwrap_or(false),
    };
    let address = input.address.filter(|a| !a.trim().is_empty());
    let location = match (address, input.lat, input.lon) {
        (Some(a), None, None) => geocode(geocoder, cache, &a, now).await?,
        (None, Some(lat), Some(lon)) => {
            if !(-90.0..=90.0).contains(&lat) || !(-180.0..=180.0).contains(&lon) {
                return Err(StartError::InvalidRequest("lat/lon out of range".into()));
            }
            Location {
                lat,
                lon,
                display_name: format!("{lat:.5}, {lon:.5}"),
            }
        }
        _ => {
            return Err(StartError::InvalidRequest(
                "give exactly one of address or lat+lon".into(),
            ));
        }
    };
    Ok(PickSession::new(pick_id, now, request, location))
}

async fn geocode<G, C>(
    geocoder: &G,
    cache: &C,
    address: &str,
    now: DateTime<Utc>,
) -> Result<Location, StartError>
where
    G: Geocoder + Sync,
    C: GeocodeCache + Sync,
{
    let key = normalise_address(address);
    if let Ok(Some(c)) = cache.get_geocode(&key).await
        && is_fresh(c.created_at, GEOCODE_TTL, now)
    {
        return Ok(c.location);
    }
    let location = geocoder
        .geocode(address)
        .await
        .map_err(|e| StartError::GeocoderUnavailable(e.to_string()))?
        .ok_or(StartError::AddressNotFound)?;
    let cached = CachedLocation {
        location: location.clone(),
        created_at: now,
    };
    if let Err(e) = cache.put_geocode(&key, &cached).await {
        tracing::warn!(error = %e, "geocode cache write failed");
    }
    Ok(location)
}

/// A new, time-ordered pick id (§1).
pub fn new_pick_id() -> String {
    ulid::Ulid::generate().to_string()
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicUsize, Ordering};

    use domain::geo::GeocodeError;
    use fat_horses_store::MemoryStore;

    use super::*;

    struct FakeGeocoder {
        calls: AtomicUsize,
    }

    impl Geocoder for FakeGeocoder {
        async fn geocode(&self, address: &str) -> Result<Option<Location>, GeocodeError> {
            self.calls.fetch_add(1, Ordering::SeqCst);
            match address {
                "down" => Err(GeocodeError::Unavailable("503".into())),
                a if a.contains("nowhere") => Ok(None),
                _ => Ok(Some(Location {
                    lat: -36.8485,
                    lon: 174.7622,
                    display_name: "Sky Tower, Auckland".into(),
                })),
            }
        }
    }

    fn geocoder() -> FakeGeocoder {
        FakeGeocoder {
            calls: AtomicUsize::new(0),
        }
    }

    fn now() -> DateTime<Utc> {
        "2026-09-21T10:00:00Z".parse().unwrap()
    }

    fn address(a: &str) -> StartInput {
        StartInput {
            address: Some(a.into()),
            ..Default::default()
        }
    }

    #[tokio::test]
    async fn defaults_and_geocoding() {
        let s = start_pick(
            &geocoder(),
            &MemoryStore::new(),
            address("Sky Tower"),
            "p1".into(),
            now(),
        )
        .await
        .unwrap();
        assert_eq!(s.request.radius_m, 200);
        assert_eq!(s.request.min_population, 10_000_000);
        assert!(!s.request.include_visited);
        assert_eq!(s.location.display_name, "Sky Tower, Auckland");
        assert_eq!(s.pick_id, "p1");
    }

    #[tokio::test]
    async fn geocodes_are_cached_by_normalised_address() {
        let g = geocoder();
        let cache = MemoryStore::new();
        for a in ["Sky Tower", "  sky   TOWER "] {
            start_pick(&g, &cache, address(a), "p".into(), now())
                .await
                .unwrap();
        }
        assert_eq!(g.calls.load(Ordering::SeqCst), 1);
    }

    #[tokio::test]
    async fn coordinates_skip_geocoding() {
        let g = geocoder();
        let input = StartInput {
            lat: Some(-36.8),
            lon: Some(174.7),
            radius_m: Some(500),
            ..Default::default()
        };
        let s = start_pick(&g, &MemoryStore::new(), input, "p".into(), now())
            .await
            .unwrap();
        assert_eq!(s.location.lat, -36.8);
        assert_eq!(s.request.radius_m, 500);
        assert_eq!(g.calls.load(Ordering::SeqCst), 0);
    }

    #[tokio::test]
    async fn validation_errors() {
        let bad = [
            StartInput::default(),
            StartInput {
                address: Some("x".into()),
                lat: Some(1.0),
                lon: Some(1.0),
                ..Default::default()
            },
            StartInput {
                lat: Some(1.0),
                ..Default::default()
            },
            StartInput {
                lat: Some(91.0),
                lon: Some(0.0),
                ..Default::default()
            },
            StartInput {
                address: Some("Sky Tower".into()),
                radius_m: Some(49),
                ..Default::default()
            },
            StartInput {
                address: Some("Sky Tower".into()),
                radius_m: Some(2001),
                ..Default::default()
            },
        ];
        for input in bad {
            let r = start_pick(
                &geocoder(),
                &MemoryStore::new(),
                input.clone(),
                "p".into(),
                now(),
            )
            .await;
            assert!(
                matches!(r, Err(StartError::InvalidRequest(_))),
                "{input:?}: {r:?}"
            );
        }
    }

    #[tokio::test]
    async fn unknown_address_and_outage() {
        let r = start_pick(
            &geocoder(),
            &MemoryStore::new(),
            address("nowhere street"),
            "p".into(),
            now(),
        )
        .await;
        assert_eq!(r.unwrap_err(), StartError::AddressNotFound);
        let r = start_pick(
            &geocoder(),
            &MemoryStore::new(),
            address("down"),
            "p".into(),
            now(),
        )
        .await;
        assert!(matches!(r, Err(StartError::GeocoderUnavailable(_))));
    }

    #[test]
    fn pick_ids_are_ulids() {
        let id = new_pick_id();
        assert_eq!(id.len(), 26);
        assert_ne!(id, new_pick_id());
    }
}
